src/invert/laplace/impls/pcr/pcr.cxx

```diff
@@ -0,0 +1,1137 @@
+  }
+
+  // Perform parallel cyclic reduction
+  cr_forward_multiple_row(aa,bb,cc,r);
+  pcr_forward_single_row(aa,bb,cc,r,x);     // Including 2x2 solver
+  cr_backward_multiple_row(aa,bb,cc,r,x);
+  // End parallel cyclic reduction
+
+  // Copy solution back to bout format - this is correct on interior rows, but
+  // not boundary rows
+  for(int kz=0; kz<nsys; kz++){
+    for(int ix=0; ix<nx; ix++){
+      x_mpi(kz,ix+xstart-xs) = x(kz,ix+1);
+    }
+  }
+
+  // Ensure solution is also correct on boundary rows
+  apply_boundary_conditions(a_mpi, b_mpi, c_mpi, r_mpi, x_mpi);
+
+}
+
+/** 
+ * Eliminate boundary rows - perform row elimination to uncouple the first and
+ * last interior rows from their respective boundary rows. This is necessary
+ * to ensure we pass a square system of interior rows to the PCR library.
+*/
+void LaplacePCR :: eliminate_boundary_rows(const Matrix<dcomplex> &a, Matrix<dcomplex> &b, const Matrix<dcomplex> &c, Matrix<dcomplex> &r) {
+
+  if (localmesh->firstX()) {
+    // x index is first interior row
+    const int xstart = localmesh->xstart;
+    for (int kz = 0; kz < nsys; kz++) {
+      b(kz,xstart) = b(kz,xstart) - c(kz, xstart-1) * a(kz,xstart) / b(kz, xstart-1);
+      r(kz,xstart) = r(kz,xstart) - r(kz, xstart-1) * a(kz,xstart) / b(kz, xstart-1);
+      // Row elimination would set a to zero, but value is unused:
+      // a(kz,xstart) = 0.0;
+    }
+  }
+  if (localmesh->lastX()) {
+    int n = xe - xs + 1; // actual length of array
+    int xind = n - localmesh->xstart - 1;
+    for (int kz = 0; kz < nsys; kz++) {
+      // x index is last interior row
+      b(kz,xind) = b(kz,xind) - c(kz, xind) * a(kz,xind+1) / b(kz, xind+1);
+      r(kz,xind) = r(kz,xind) - c(kz, xind) * r(kz,xind+1) / b(kz, xind+1);
+      // Row elimination would set c to zero, but value is unused:
+      // c(kz,xind) = 0.0;
+    }
+  }
+}
+
+/** 
+ * Apply the boundary conditions on the first and last X processors
+*/
+void LaplacePCR :: apply_boundary_conditions(const Matrix<dcomplex> &a, const Matrix<dcomplex> &b, const Matrix<dcomplex> &c, const Matrix<dcomplex> &r,Matrix<dcomplex> &x) {
+
+  if (localmesh->firstX()) {
+    for (int kz = 0; kz < nsys; kz++) {
+      for(int ix = localmesh->xstart-1; ix >= 0; ix--){
+	x(kz,ix) = (r(kz, ix) - c(kz, ix) * x(kz,ix+1)) / b(kz, ix);
+      }
+    }
+  }
+  if (localmesh->lastX()) {
+    int n = xe - xs + 1; // actual length of array
+    for (int kz = 0; kz < nsys; kz++) {
+      for(int ix = n - localmesh->xstart; ix < n; ix++){
+	x(kz,ix) = (r(kz,ix) - a(kz, ix) * x(kz,ix-1)) / b(kz, ix);
+      }
+    }
+  }
+}
+
+/** 
+ * @brief   Forward elimination of CR until a single row per MPI process remains.
+ * @details After a single row per MPI process remains, PCR or CR between a single row is performed.
+*/
+void LaplacePCR :: cr_forward_multiple_row(Matrix<dcomplex> &a,Matrix<dcomplex> &b,Matrix<dcomplex> &c,Matrix<dcomplex> &r)
+{
+    MPI_Comm comm = BoutComm::get();
+    int i, l;
```

Comment:
Clang-tidy will probably complain about these, but I think these declarations can be moved into the loop init statements